src/charm.py

```diff
@@ -0,0 +1,595 @@
+
+logger = logging.getLogger(__name__)
+
+
+class ReconcileRunnersEvent(EventBase):
+    """Event representing a periodic check to ensure runners are ok."""
+
+
+class UpdateRunnerBinEvent(EventBase):
+    """Event representing a periodic check for new versions of the runner binary."""
+
+
+CharmT = TypeVar("CharmT")
+EventT = TypeVar("EventT")
+
+
+def catch_unexpected_charm_errors(
+    func: Callable[[CharmT, EventT], None]
+) -> Callable[[CharmT, EventT], None]:
+    """Catch unexpected errors in charm.
+
+    This decorator is for unrecoverable errors and sets the charm to
+    `BlockedStatus`.
+
+    Args:
+        func: Charm function to be decorated.
+
+    Returns:
+        Decorated charm function with catching unexpected errors.
+    """
+
+    @functools.wraps(func)
+    def func_with_catch_unexpected_errors(self, event: EventT) -> None:
+        # Safe guard against unexpected error.
+        try:
+            func(self, event)
+        except Exception as err:  # pylint: disable=broad-exception-caught
+            logger.exception(err)
+            self.unit.status = BlockedStatus(str(err))
+
+    return func_with_catch_unexpected_errors
+
+
+def catch_unexpected_action_errors(
+    func: Callable[[CharmT, ActionEvent], None]
+) -> Callable[[CharmT, ActionEvent], None]:
+    """Catch unexpected errors in actions.
+
+    Args:
+        func: Action function to be decorated.
+
+    Returns:
+        Decorated charm function with catching unexpected errors.
+    """
+
+    @functools.wraps(func)
+    def func_with_catch_unexpected_errors(self, event: ActionEvent) -> None:
+        # Safe guard against unexpected error.
+        try:
+            func(self, event)
+        except Exception as err:  # pylint: disable=broad-exception-caught
+            logger.exception(err)
+            event.fail(f"Failed to get runner info: {err}")
+
+    return func_with_catch_unexpected_errors
+
+
+class GithubRunnerCharm(CharmBase):
+    """Charm for managing GitHub self-hosted runners."""
+
+    _stored = StoredState()
+
+    service_token_path = Path("service_token")
+    repo_check_web_service_path = Path("/home/ubuntu/repo_policy_compliance_service")
+    repo_check_web_service_script = Path("src/repo_policy_compliance_service.py")
+    repo_check_systemd_service = Path("/etc/systemd/system/repo-policy-compliance.service")
+
+    def __init__(self, *args, **kargs) -> None:
+        """Construct the charm.
+
+        Args:
+            args: List of arguments to be passed to the `CharmBase` class.
+            kargs: List of keyword arguments to be passed to the `CharmBase`
+                class.
+        """
+        super().__init__(*args, **kargs)
+
+        self._event_timer = EventTimer(self.unit.name)
+
+        self._stored.set_default(
+            path=self.config["path"],  # for detecting changes
+            runner_bin_url=None,
+        )
+
+        self.proxies: ProxySetting = {}
```

Comment:
This looks like it should be charm state

Reply:
Minor comments